### ISSUE TITLE
Aggregate releases

### ISSUE THREAD

**Issue by Andarist** [atlassian/changesets#683]
It's a very popular request to aggregate all of the changelogs stuff in a single GitHub release. A lot of open-source monorepos revolve around a single project and thus creating dozens of releases with each `changeset publish` gets quite noisy - this is especially true for linked packages or fixed packages (FYI I'm working on introducing the fixed package groups).

I propose to introduce support for this:
```td
type LinkedPatterns = string[];
type Linked = Array<
  | { name?: string; packages: LinkedPatterns; aggregate?: boolean }
  | LinkedPatterns
>;
```

Since the goal of this feature is to combine things into a single GitHub release we need something to put in the title of the release. The `name` option would be used for that and the created release would have such title `${name} v${version}`. It IMHO makes sense to infer the name from the used scope as that would work for the majority of use cases - but that can also be left out from the MVP. 

`aggregate` (or `aggregateRelease`) would be an opt-in for this feature - `@changesets/action` could pick this up and decide how to handle the releasing part of the things. Some people have also suggested that a single CHANGELOG could be created (that would be a copy of the aggregated release) - I think there is a value in this feature, but I'm not sure if this should automatically be enabled with just `aggregate` or if this requires an additional option. I'm afraid of introducing an option creep though so I would really prefer to keep this simple. 

The big question is - how do we concatenate particular changelogs? How do we sort packages? My ideas:
- allow for stuff defined in `packages` to decide about this, for instance: `packages: ['@chakra-ui/react', '@chakra-ui/*']` would tell Changesets to use the current CHANGELOG update from `@chakra-ui/react` first and to put everything else after that
- we could try to sort the packages based on the dependency graph, so basically try to infer the order from the graph - the packages with the most descendants would take precedence over packages with less of those. This is kinda magical and probably just a possible enhancement over the previous idea. I would keep this one out of the MVP

There is also an additional problem - some projects, like Chakra, have an "aggregate" package like `@chakra-ui/react`. The purpose of this package is to just reexport stuff from everything else and nothing beyond that. I think they currently just add `@chakra-ui/react` to each created changeset because of that. This would mean that every change in the aggregate release would get kinda duplicated because of that - as it would be described both by `@chakra-ui/react` and some other, inner, package like `@chakra-ui/button` or something. Could their workflow be adjusted with the `aggregate` in place in a way that they wouldn't suffer from this problem?

Do we want to use this in Changesets itself? It feels like we fall into the "single project" category - but we are not using linked nor fixed packages at the moment. And I think it makes sense for us to keep versioning for internal packages separate from the `@changesets/cli` version - because we tend to introduce subtle changes there that could be seen as breaking ones. With how we are handling this today we have the freedom of creating major bumps for those packages as frequently as we want. Could we somehow allow `aggregate` to work in monorepos without linked or fixed packages?

@ianstormtaylor @with-heart

**Comment by NMinhNguyen** [comment]
Would also like to mention Babel's changelog format: https://github.com/changesets/changesets/pull/537#issuecomment-968407244

**Comment by Andarist** [comment]
I'm thinking about a similar format to that - I'm unsure though if we are going to be able to correlate packages to particular changes. I was imagining just flattening the whole thing, introducing "names" for package groups (linked or fixed) and treating them as "pseudopackages". So instead of writing a Changeset like this:
```yml
---
'@changesets/cli': minor
'@changesets/assemble-release-plan': patch
---

Content
```
You would write this:
```yml
---
'changesets': minor
---

Content
```

However, this would not play that well with linked packages - if we could only use the group name here then all linked packages would get always released. This would effectively make them behave like fixed packages.

So perhaps, we shouldn't treat those names as pseudo packages :thinking: I'll have to think about it some more - but any comments from the community would be highly appreciated here.

**Comment by paales** [comment]
@Andarist We're very interested in these aggregated releases

Our current release notes are practically unusable on GitHub because we get a lot of duplicates which makes reading the [release notes](https://github.com/graphcommerce-org/graphcommerce/releases) impossible.

We've got lots of duplicates for example:

- https://github.com/graphcommerce-org/graphcommerce/releases/tag/%40graphcommerce%2Fnext-ui%404.7.2
- https://github.com/graphcommerce-org/graphcommerce/releases/tag/%40graphcommerce%2Fmollie-magento-payment%403.2.9
- And a ton more

In an ideal scenario we'd like to:

- Keep the separate versioning of the individual packages because we have multiple 'groups' of packages. We are considering linking those groups together based on a wildcard, but we will always have multiple groups.

- The 'root' ([`@graphcommerce/graphcommerce`](https://github.com/graphcommerce-org/graphcommerce/blob/main/package.json)) should catch all changes of all child packages and have the release notes aggregated there. This may also be versioned, but I don't reallly care :) 

- We don't want to have separate releases on GitHub for separate packages as this just isn't clear. We only want to create a GitHub release for the root package, all other packages can have their CHANGELOG.md and their git tag of course, but not a GitHub release note.

Anything I can help with to move this forward?

**Comment by Andarist** [comment]
If you'd be willing to implement this - I would be more than happy to dedicate my time to talk through the design of this feature, review a PR etc.

**Comment by thdxr** [comment]
We're also interested in this feature at https://github.com/serverless-stack/serverless-stack 

We currently have a set of `fixed` packages that are always bumped together.  When we do a release we'd like a *single* tag and a *single* release published with the list of changes

Definitely willing to submit a PR if I could understand better of how this should be designed

**Comment by dotansimha** [comment]
Also super interested about this. 
We are having projects like https://github.com/dotansimha/graphql-code-generator , and the Releases tab is bloated with tons of releases. 
I was thinking of an option to allow to group all changes in a single release, and publish a summary with a list of all modified packages.

**Comment by mblackrittr** [comment]
Yeah, as much as I favor Changesets (atop of Conventional Commits), aggregation is *THE* missing feature - #OhHellboyYeah for sure! 😻

Because I was already dreaming bad about all the release bloat it'll push straight from hell! 🙀
